### PR TITLE
fix(recovery): :bug: false-positive for non-gateway URLs

### PR DESCRIPTION
**Pull request by whizzzkid**
Closes: #1162 

This is a case I should've thought of when working on: #1125. The reason this happens:

- When `Automatic Mode` is turned off, the redirect actually happens before the recovery mode begins.
- Since the redirect happens to local node, the recovery mode gets triggered, asking to redirect to public gateway.
- The public gateway redirects back to local and results in the endless loop.

Fix: <del>Add check for `!state.redirect`</del> remove false-positives from `sameGateway` check.

I would like to merge #1160 and #1161 before this and create a patch release.

### CODE REVIEW

File: add-on/src/lib/ipfs-request.js [hunk 1]
```diff
@@ -144,7 +144,11 @@ export function createRequestModifier (getState, dnslinkResolver, ipfsPathValida
 
       // When local IPFS node is unreachable , show recovery page where user can redirect
       // to public gateway.
-      if (!state.nodeActive && request.type === 'main_frame' && sameGateway(request.url, state.gwURL)) {
+      if (!state.nodeActive &&
+        !state.redirect &&
+        request.type === 'main_frame' &&
+        sameGateway(request.url, state.gwURL)
+      ) {
```
Comment (autonome) [inline add-on/src/lib/ipfs-request.js:151]: Please add a comment for each of the cases tested in that statement, so future generations can understand what's going on and why.

File: add-on/src/lib/ipfs-request.js [hunk 2]
```diff
@@ -144,7 +144,11 @@ export function createRequestModifier (getState, dnslinkResolver, ipfsPathValida
 
       // When local IPFS node is unreachable , show recovery page where user can redirect
       // to public gateway.
-      if (!state.nodeActive && request.type === 'main_frame' && sameGateway(request.url, state.gwURL)) {
+      if (!state.nodeActive && //                  node is not active
+        !state.redirect && //                      this is not a redirect request
```
Comment (lidel) [inline add-on/src/lib/ipfs-request.js]: nit: this is a global flag that controls if companion is redirecting requests to local gateway, or not.
This was not fixing the problem, and the comment was only confusing the reader – I've restored old version and applied proper fix in `sameGateway`.

**Review by lidel — CHANGES_REQUESTED** [review]
The problem with Companion is that you need to think with portals ;-) 

I dont believe this is fixing problem from https://twitter.com/unicomp21/status/1626244123102679041

We can clearly detect when someone opens URL equal to the Gateway `http://localhost:8080/` – I'll push fix shortly.

File: test/functional/lib/ipfs-path.test.js [hunk 3]
```diff
@@ -184,11 +184,23 @@ describe('ipfs-path.js', function () {
       const gw = 'http://localhost:8080'
       expect(sameGateway(url, gw)).to.equal(true)
     })
-    it('should return true on 127.0.0.1/0.0.0.0 host match', function () {
+    it('should return true on localhost subdomain host match', function () {
+      const url = 'http://bafybeigdyrzt5sfp7udm7hu76uh7y26nf3efuylqabf3oclgtqy55fbzdi.ipfs.localhost:8080/foo/bar'
```
Comment (lidel) [inline test/functional/lib/ipfs-path.test.js:188]: :information_source: we were missing subdomain test, so I've added it

File: test/functional/lib/ipfs-path.test.js [hunk 4]
```diff
@@ -184,11 +184,23 @@ describe('ipfs-path.js', function () {
...
+    it('should return true on RPC webui path 127.0.0.1/0.0.0.0 host match', function () {
+      // this is misconfiguration, but handling it in sameGateway ensures users who do this dont waste our time debugging
       const url = 'http://0.0.0.0:5001/webui'
       const api = 'http://127.0.0.1:5001'
       expect(sameGateway(url, api)).to.equal(true)
     })
+    it('should return true on RPC /api/v0 path 127.0.0.1/0.0.0.0 host match', function () {
+      // this is misconfiguration, but handling it in sameGateway ensures users who do this dont waste our time debugging
+      const url = 'http://0.0.0.0:5001/api/v0/id'
+      const api = 'http://127.0.0.1:5001'
```
Comment (lidel) [inline test/functional/lib/ipfs-path.test.js:201]: :information_source:  this saves us a day of debugging when user puts wrong port in wrong field in their config, and then tweets about it ;-)

File: test/functional/lib/ipfs-request-gateway-redirect.test.js [hunk 5]
```diff
@@ -431,17 +431,38 @@ describe('modifyRequest.onBeforeRequest:', function () {
...
+      const request = url2request('https://localhost:8080/')
+      expect(modifyRequest.onBeforeRequest(request)).to.equal(undefined)
+    })
+    it('should not show recovery page if extension is disabled', function () {
+      // allows user to quickly avoid anything similar to https://github.com/ipfs/ipfs-companion/issues/1162
+      state.active = false
+      expect(state.nodeActive).to.be.equal(false)
+      const request = url2request('https://localhost:8080/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR/foo/bar')
+      expect(modifyRequest.onBeforeRequest(request)).to.equal(undefined)
+    })
```
Comment (lidel) [inline test/functional/lib/ipfs-request-gateway-redirect.test.js:465]: :information_source:  I think this covers all edge cases, and ensures recovery logic does not depend on redirect state (which was why it was initially difficult to reproduce the bug from https://github.com/ipfs/ipfs-companion/issues/1162)

**Review by lidel — APPROVED** [review]
@whizzzkid pushed alternative fix for `sameGateway` check and some tests.
This approach should be more reliable.

File: add-on/src/lib/ipfs-request.js [hunk 6]
```diff
@@ -144,7 +144,11 @@ export function createRequestModifier (getState, dnslinkResolver, ipfsPathValida
...
+        request.type === 'main_frame' && //        this is a request for a root document
+        sameGateway(request.url, state.gwURL) //   this is a request to the local gateway
```
[the remark this answers is not in the document]
Comment (lidel) [reply add-on/src/lib/ipfs-request.js]: `state` is global, not per request. it informs redirect logic what are current settings and state of RPC. 
(bit of a legacy thing, from the XUL time (before v2) when it was less expensive to cache options in one place, but it is still useful when writing tests, to simulate specific state of world).

you probably saw us setting it in tests per different requests to simultate specific  flows, and assumed it is per request, but it is is not :)
